File: src/stores/account.ts
import { defineStore } from "pinia";
import { ref, type Ref } from "vue";
import { tauriCommand, type CommandError } from "@/composables/useTauriCommand";
import type {
  Account,
  DeviceCodeResponse,
  LoginState,
} from "@/types/account";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export const useAccountStore = defineStore("account", () => {
  // ── State ───────────────────────────────────────────────
  const accounts: Ref<Account[]> = ref([]);
  const activeAccount: Ref<Account | null> = ref(null);
  const loginState: Ref<LoginState> = ref('idle');
  const deviceCode: Ref<DeviceCodeResponse | null> = ref(null);
  const error: Ref<string | null> = ref(null);

  /** Unlisteners for Tauri events. */
  const unlisteners: UnlistenFn[] = [];

  // ── Actions ────────────────────────────────────────────

  /** Load all accounts from the backend. */
  async function loadAccounts(): Promise<void> {
    try {
      // 命令名与 Rust #[tauri::command] fn 名一致：get_accounts
      accounts.value = await tauriCommand<Account[]>("get_accounts");
      activeAccount.value = accounts.value.find((a) => a.is_active) ?? null;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
    }
  }

  /** Start Microsoft login via device code flow. */
  async function loginMicrosoft(): Promise<void> {
    loginState.value = 'pending';
    error.value = null;
    deviceCode.value = null;
    try {
      // login_microsoft 是一个完整的异步流程（设备码轮询）
      // Rust 后端通过事件推送设备码，命令完成后返回账号
      // 先注册事件监听，再调用命令
      const unlisten = await listen<DeviceCodeResponse>("account:device_code", (event) => {
        deviceCode.value = event.payload;
        loginState.value = 'awaiting_user';
      });
      unlisteners.push(unlisten);

      // login_microsoft 会阻塞直到用户完成授权（最长5分钟）
      const account = await tauriCommand<Account>("login_microsoft");
      accounts.value.push(account);
      activeAccount.value = account;
      loginState.value = 'success';
      deviceCode.value = null;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
      loginState.value = 'error';
      deviceCode.value = null;
    }
  }

  /** Login with an offline (cracked) account. */
  async function loginOffline(username: string): Promise<Account | null> {
    loginState.value = 'pending';
    error.value = null;
    try {
      const account = await tauriCommand<Account>("login_offline", { username });
      accounts.value.push(account);
      activeAccount.value = account;
      loginState.value = 'success';
      return account;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
      loginState.value = 'error';
      return null;
    }
  }

  /** Logout and remove an account. */
  async function logout(id: string): Promise<boolean> {
    try {
      await tauriCommand<void>("logout", { id });
      accounts.value = accounts.value.filter((a) => a.id !== id);
      if (activeAccount.value?.id === id) {
        activeAccount.value = accounts.value.find((a) => a.is_active) ?? null;
      }
      return true;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
      return false;
    }
  }

  /** Set the active account. */
  async function setActive(id: string): Promise<boolean> {
    try {
      await tauriCommand<void>("set_active_account", { id });
      const account = accounts.value.find((a) => a.id === id);
      if (account) {
        // Update is_active flags locally
        for (const a of accounts.value) {
          a.is_active = a.id === id;
        }
        activeAccount.value = account;
      }
      return true;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
      return false;
    }
  }

  /** Refresh an account's access token. */
  async function refreshAccount(id: string): Promise<boolean> {
    try {
      const updated = await tauriCommand<Account>("refresh_account", { id });
      const idx = accounts.value.findIndex((a) => a.id === updated.id);
      if (idx >= 0) {
        accounts.value[idx] = updated;
      }
      if (activeAccount.value?.id === updated.id) {
        activeAccount.value = updated;
      }
      return true;
    } catch (e) {
      const cmdErr = e as CommandError;
      error.value = cmdErr.message;
      return false;
    }
  }

  /** Reset login state. */
  function resetLoginState(): void {
    loginState.value = 'idle';
    deviceCode.value = null;
    error.value = null;
  }

  /** Cleanup event listeners. */
  function cleanup(): void {
    for (const unlisten of unlisteners) {
      unlisten();
    }
    unlisteners.length = 0;
  }

  return {
    accounts,
    activeAccount,
    loginState,
    deviceCode,
    error,
    loadAccounts,
    loginMicrosoft,
    loginOffline,
    logout,
    setActive,
    refreshAccount,
    resetLoginState,
    cleanup,
  };
});

File: src-tauri/src/commands/account.rs
//! Account-related Tauri commands.
//!
//! 提供微软登录、离线登录、账号管理相关的 IPC 命令。

use crate::error::{CommandResponse, CommandResult};
use crate::models::account::Account;
use crate::services::account_service;
use crate::state::AppState;
use crate::utils::db_pool;
use tauri::State;

/// 微软登录（OAuth 2.0 设备代码流）。
///
/// 启动微软登录流程，完成后返回创建的账号信息。
#[tauri::command]
pub async fn login_microsoft(
    state: State<'_, AppState>,
) -> CommandResult<Account> {
    let account = account_service::login_microsoft(&state).await?;
    Ok(CommandResponse::ok(account))
}

/// 离线登录。
///
/// 使用自定义用户名创建离线账号，无需网络认证。
#[tauri::command]
pub async fn login_offline(
    state: State<'_, AppState>,
    username: String,
) -> CommandResult<Account> {
    let pool = db_pool::get_pool(&state)?;
    let account = account_service::login_offline(pool, &username).await?;
    Ok(CommandResponse::ok(account))
}

/// 获取所有账号列表。
///
/// 返回所有已登录的账号，按激活状态和创建时间排序。
#[tauri::command]
pub async fn get_accounts(
    state: State<'_, AppState>,
) -> CommandResult<Vec<Account>> {
    let pool = db_pool::get_pool(&state)?;
    let accounts = account_service::get_accounts(pool).await?;
    Ok(CommandResponse::ok(accounts))
}

/// 获取当前激活的账号。
///
/// 返回当前正在使用的账号信息。如果没有激活账号则返回错误。
#[tauri::command]
pub async fn get_active_account(
    state: State<'_, AppState>,
) -> CommandResult<Account> {
    let pool = db_pool::get_pool(&state)?;
    let account = account_service::get_active_account(pool).await?;
    Ok(CommandResponse::ok(account))
}

/// 设置激活账号。
///
/// 将指定 ID 的账号设为激活状态，其他账号自动取消激活。
#[tauri::command]
pub async fn set_active_account(
    state: State<'_, AppState>,
    id: String,
) -> CommandResult<Account> {
    let pool = db_pool::get_pool(&state)?;
    let account = account_service::set_active_account(pool, &id).await?;
    Ok(CommandResponse::ok(account))
}

/// 登出并删除账号。
///
/// 删除指定 ID 的账号，同时清除 OS 密钥链中的令牌。
#[tauri::command]
pub async fn logout(
    state: State<'_, AppState>,
    id: String,
) -> CommandResult<()> {
    let pool = db_pool::get_pool(&state)?;
    account_service::logout(pool, &id).await?;
    Ok(CommandResponse::ok(()))
}

/// 刷新账号令牌。
///
/// 仅支持微软账号。使用存储的 refresh token 重新认证整个链路。
#[tauri::command]
pub async fn refresh_account(
    state: State<'_, AppState>,
    id: String,
) -> CommandResult<Account> {
    let account = account_service::refresh_account(&state, &id).await?;
    Ok(CommandResponse::ok(account))
}
